File: app/routers/transactions.py
"""
Transactions router.

POST   /transactions              initiate a transaction (customer)
GET    /transactions              list my transactions
GET    /transactions/:id          get one transaction
POST   /transactions/:id/confirm  provider confirms
POST   /transactions/:id/rate     customer rates provider
"""

import uuid
from typing import Optional

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.security import get_current_user
from app.database import get_db
from app.schemas.transaction import (
    RatingCreateIn,
    RatingOut,
    TransactionCreateIn,
    TransactionListOut,
    TransactionOut,
)
from app.services import transaction_service

router = APIRouter()


# ── POST /transactions ────────────────────────────────────────────────────────
@router.post(
    "",
    response_model=TransactionOut,
    status_code=status.HTTP_201_CREATED,
    summary="Initiate a transaction",
    description=(
        "Customer initiates a transaction with a provider. "
        "The customer auto-confirms their side. "
        "The transaction stays 'pending' until the provider also confirms."
    ),
)
def initiate_transaction(
    body: TransactionCreateIn,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return transaction_service.initiate_transaction(
        data=body, current_user=current_user, db=db
    )


# ── GET /transactions ─────────────────────────────────────────────────────────
@router.get(
    "",
    response_model=TransactionListOut,
    status_code=status.HTTP_200_OK,
    summary="List my transactions",
    description="Returns all transactions where the authenticated user is provider or customer.",
)
def list_transactions(
    status: Optional[str] = Query(None, description="Filter: pending | confirmed | expired | disputed"),
    page:      int = Query(1,  ge=1),
    page_size: int = Query(20, ge=1, le=50),
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return transaction_service.list_transactions(
        current_user=current_user,
        db=db,
        status_filter=status,
        page=page,
        page_size=page_size,
    )


# ── GET /transactions/:id ─────────────────────────────────────────────────────
@router.get(
    "/{tx_id}",
    response_model=TransactionOut,
    status_code=status.HTTP_200_OK,
    summary="Get a transaction",
)
def get_transaction(
    tx_id: uuid.UUID,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return transaction_service.get_transaction(
        tx_id=tx_id, current_user=current_user, db=db
    )


# ── POST /transactions/:id/confirm ────────────────────────────────────────────
@router.post(
    "/{tx_id}/confirm",
    response_model=TransactionOut,
    status_code=status.HTTP_200_OK,
    summary="Confirm a transaction",
    description=(
        "Either the provider or the customer can call this. "
        "When both have confirmed, the transaction becomes 'confirmed' (immutable)."
    ),
)
def confirm_transaction(
    tx_id: uuid.UUID,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return transaction_service.confirm_transaction(
        tx_id=tx_id, current_user=current_user, db=db
    )


# ── POST /transactions/:id/rate ───────────────────────────────────────────────
@router.post(
    "/{tx_id}/rate",
    response_model=RatingOut,
    status_code=status.HTTP_201_CREATED,
    summary="Rate a transaction",
    description=(
        "Customer submits a thumbs-up or thumbs-down rating. "
        "Can be submitted as soon as the customer has confirmed their side. "
        "One rating per transaction — immutable once submitted."
    ),
)
def rate_transaction(
    tx_id: uuid.UUID,
    body: RatingCreateIn,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return transaction_service.rate_transaction(
        tx_id=tx_id, data=body, current_user=current_user, db=db
    )

File: mobile/services/transactions.ts
/**
 * services/transactions.ts
 *
 * Wired to the real FastAPI endpoints:
 *
 *   POST /transactions               initiate
 *   GET  /transactions               list mine
 *   GET  /transactions/:id           get one
 *   POST /transactions/:id/confirm   provider confirms
 *   POST /transactions/:id/rate      customer rates
 */

import { api } from "./api";

export interface RatingOut {
  id: string;
  thumbs_up: boolean;
  comment: string | null;
  created_at: string;
}

export interface TransactionOut {
  id: string;
  provider_id: string;
  customer_id: string;
  amount_xaf: number;
  description: string | null;
  status: "pending" | "confirmed" | "expired" | "disputed";
  initiated_by: string;
  initiated_at: string;
  expires_at: string;
  provider_confirmed_at: string | null;
  customer_confirmed_at: string | null;
  provider_name: string;
  customer_name: string | null;
  rating: RatingOut | null;
}

export interface TransactionListOut {
  total: number;
  page: number;
  page_size: number;
  results: TransactionOut[];
}

export async function initiateTransaction(payload: {
  provider_id: string;
  amount_xaf: number;
  description?: string;
}): Promise<TransactionOut> {
  return api.post<TransactionOut>("/transactions", payload);
}

export async function listTransactions(params: {
  status?: string;
  page?: number;
  page_size?: number;
}): Promise<TransactionListOut> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  qs.set("page",      String(params.page      ?? 1));
  qs.set("page_size", String(params.page_size ?? 20));
  return api.get<TransactionListOut>(`/transactions?${qs.toString()}`);
}

export async function getTransaction(id: string): Promise<TransactionOut> {
  return api.get<TransactionOut>(`/transactions/${id}`);
}

export async function confirmTransaction(id: string): Promise<TransactionOut> {
  return api.post<TransactionOut>(`/transactions/${id}/confirm`, {});
}

export async function rateTransaction(
  id: string,
  payload: { thumbs_up: boolean; comment?: string },
): Promise<RatingOut> {
  return api.post<RatingOut>(`/transactions/${id}/rate`, payload);
}
